<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Example multiple sequences data — example_many_sequences • ggDNAvis</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Example multiple sequences data — example_many_sequences"><meta name="description" content="A collection of made-up sequences in the style of long reads over a repeat region
(e.g. NOTCH2NLC), with meta-data describing the participant each read is from and
the family each participant is from. Can be used in visualise_many_sequences(),
visualise_methylation(), and helper functions to visualise these sequences.
Generation code is available at data-raw/example_many_sequences.R"><meta property="og:description" content="A collection of made-up sequences in the style of long reads over a repeat region
(e.g. NOTCH2NLC), with meta-data describing the participant each read is from and
the family each participant is from. Can be used in visualise_many_sequences(),
visualise_methylation(), and helper functions to visualise these sequences.
Generation code is available at data-raw/example_many_sequences.R"><meta property="og:image" content="https://ejade42.github.io/ggDNAvis/logo.png"></head><body><p>









  pkgdown/google0acd00d8574ec2df.html

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggDNAvis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ejade42/ggDNAvis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Example multiple sequences data</h1>
      <small class="dont-index">Source: <a href="https://github.com/ejade42/ggDNAvis/blob/main/R/ggDNAvis-package.R" class="external-link"><code>R/ggDNAvis-package.R</code></a></small>
      <div class="d-none name"><code>example_many_sequences.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A collection of made-up sequences in the style of long reads over a repeat region
(e.g. <em>NOTCH2NLC</em>), with meta-data describing the participant each read is from and
the family each participant is from. Can be used in <code><a href="visualise_many_sequences.html">visualise_many_sequences()</a></code>,
<code><a href="visualise_methylation.html">visualise_methylation()</a></code>, and helper functions to visualise these sequences.<br><br>
Generation code is available at <code>data-raw/example_many_sequences.R</code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="va">example_many_sequences</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="format">Format<a class="anchor" aria-label="anchor" href="#format"></a></h2>

<div class="section">
<h3 id="example-many-sequences"><code>example_many_sequences</code><a class="anchor" aria-label="anchor" href="#example-many-sequences"></a></h3>


<p>A dataframe with 23 rows and 10 columns:</p><dl><dt>family</dt>
<dd><p>Participant family</p></dd>

<dt>individual</dt>
<dd><p>Participant ID</p></dd>

<dt>read</dt>
<dd><p>Unique read ID</p></dd>

<dt>sequence</dt>
<dd><p>DNA sequence of the read</p></dd>

<dt>sequence_length</dt>
<dd><p>Length (nucleotides) of the read</p></dd>

<dt>quality</dt>
<dd><p>FASTQ quality scores for the read. Each character represents a score from 0 to 40 - see <code><a href="fastq_quality_scores.html">fastq_quality_scores</a></code>.<br><br> These values are made up via <code>pmin(pmax(round(rnorm(n, mean = 20, sd = 10)), 0), 40)</code> i.e. sampled from a normal distribution with mean 20 and standard deviation 10, then rounded to integers between 0 and 40 (inclusive) - see <code>example_many_sequences.R</code></p></dd>

<dt>methylation_locations</dt>
<dd><p>Indices along the read (starting at 1) at which methylation probability was assessed i.e. CpG sites. Stored as a single character value per read, condensed from a numeric vector via <code><a href="vector_to_string.html">vector_to_string()</a></code>.</p></dd>

<dt>methylation_probabilities</dt>
<dd><p>Probability of methylation (8-bit integer i.e. 0-255) for each assessed base. Stored as a single character value per read, condensed from a numeric vector via <code><a href="vector_to_string.html">vector_to_string()</a></code>.<br><br> These values are made up via <code>round(runif(n, min = 0, max = 255))</code> - see <code>example_many_sequences.R</code></p></dd>

<dt>hydroxymethylation_locations</dt>
<dd><p>Indices along the read (starting at 1) at which hydroxymethylation probability was assessed i.e. CpG sites. Stored as a single character value per read, condensed from a numeric vector via <code><a href="vector_to_string.html">vector_to_string()</a></code>.</p></dd>

<dt>hydroxymethylation_probabilities</dt>
<dd><p>Probability of hydroxymethylation (8-bit integer i.e. 0-255) for each assessed base. Stored as a single character value per read, condensed from a numeric vector via <code><a href="vector_to_string.html">vector_to_string()</a></code>.<br><br> These values are made up via <code>round(runif(n, min = 0, max = 255 - this_base_methylation_probability))</code> such that the summed methylation and hydroxymethylation probability never exceeds 255 (100%) - see <code>example_many_sequences.R</code></p></dd>


</dl></div>

    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evelyn Jade.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  
</body></html>

