<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Reverse modification locations if needed (merge_methylation_with_metadata() helper) — reverse_locations_if_needed • ggDNAvis</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Reverse modification locations if needed (merge_methylation_with_metadata() helper) — reverse_locations_if_needed"><meta name="description" content="This function takes a vector of condensed modification locations/indices (e.g.
c(&quot;3,6,9,12&quot;, &quot;1,4,7,10&quot;)), a vector of directions (which must all be either
&quot;forward&quot; or &quot;reverse&quot;, not case-sensitive), and a vector of sequence lengths
(integers).
Returns a vector of condensed locations where reads that were originally forward
are unchanged, and reads that were originally reverse are flipped to now be forward.
Optionally, a numerical offset can be set. If this is left at 0 (the default value),
then a CpG assessed for methylation would be reverse-complemented to a CG with the
modification information ascribed to the G (as the G is at the location where the actual
modified C was on the other strand). However, setting the offset to 1 would shift all
of the modification indices by 1 such that the modification is now ascribed to the C of the
reverse-strand CG. This is beneficial for visualising the modifications as it ensures consistency
between originally-forward and originally-reverse strands by making the modification score associated
with each CpG site always be located at the C, but may be misleading for quantitative analysis.
Setting the offset to anything other than 0 or 1 should work but may be biologically misleading,
so produces a warning.
Called by merge_methylation_with_metadata() to create a forward dataset, alongside
reverse_sequence_if_needed(), reverse_quality_if_needed(), and reverse_probabilities_if_needed().
Example:
Forward sequence, with indices of Cs in CpGs numbered:
C C C A G G C G G C G G C G A C C G A
            7     10    13      17
(length = 19, locations = &quot;7,10,13,17&quot;, CpGs = 7-8, 10-11, 13-14, 17-18)
Reverse sequence, with indices of C in CpGs numbered:
T C G G T C G C C G C C G C C T G G G
  2       6     9     12
(length = 19, locations = &quot;2,6,9,12&quot;, CpGs = 2-3, 6-7, 9-10, 12-13)
As CG reverse-complements to itself, each CpG site has a 1:1 correspondence with
a CpG site in the reverse strand. Many methylation calling models assess C-methylation
at the C of each CpG. To map the locations from C to C, we take 19 - &amp;lt;index&amp;gt; such that
&quot;7,10,13,17&quot; becomes &quot;12,9,6,2&quot; and &quot;2,6,9,12&quot; becomes &quot;17,13,10,7&quot;.
The symmetry of CpGs means mapping from C to C is also symmetric.
This is achieved by setting offset = 1, as mapping C to C involves shifting position by 1.
Conversely, to map the locations from C to G (i.e. preserving the actual location of each
modification, which is required if assessed locations are non-symmetric/don't reverse-complement
to themselves like CpGs do), we take 20 - &amp;lt;index&amp;gt; such that
&quot;7,10,13,17&quot; becomes &quot;13,10,7,3&quot; i.e. the indices of the Gs in CpGs in the reverse
sequence. Likewise &quot;2,6,9,12&quot; becomes &quot;18,14,11,8&quot; i.e. the indices of the Gs in CpGs in
the forward sequence.
This is achieved by setting offset = 0, as mapping C to G preserves the actual original position
at which each modification was assessed, but changes the base to its complement.
In general, new locations are calculated as (&amp;lt;length&amp;gt; + 1 - &amp;lt;offset&amp;gt;) - &amp;lt;index&amp;gt;.
Of course, output locations are reversed before returning so that they all
return in ascending order, as is standard for all location vectors/strings.
If wanting to write reversed sequences to FASTQ via write_modified_fastq(), locations
must be symmetric (e.g. CpG) and offset must be set to 1. Asymmetric locations are impossible
to write to modified FASTQ once reversed because then e.g. cytosine methylation will be assessed
at guanines, which SAMtools can't account for. Symmetrically reversing CpGs via offset = 1 is
the only way to fix this."><meta property="og:description" content="This function takes a vector of condensed modification locations/indices (e.g.
c(&quot;3,6,9,12&quot;, &quot;1,4,7,10&quot;)), a vector of directions (which must all be either
&quot;forward&quot; or &quot;reverse&quot;, not case-sensitive), and a vector of sequence lengths
(integers).
Returns a vector of condensed locations where reads that were originally forward
are unchanged, and reads that were originally reverse are flipped to now be forward.
Optionally, a numerical offset can be set. If this is left at 0 (the default value),
then a CpG assessed for methylation would be reverse-complemented to a CG with the
modification information ascribed to the G (as the G is at the location where the actual
modified C was on the other strand). However, setting the offset to 1 would shift all
of the modification indices by 1 such that the modification is now ascribed to the C of the
reverse-strand CG. This is beneficial for visualising the modifications as it ensures consistency
between originally-forward and originally-reverse strands by making the modification score associated
with each CpG site always be located at the C, but may be misleading for quantitative analysis.
Setting the offset to anything other than 0 or 1 should work but may be biologically misleading,
so produces a warning.
Called by merge_methylation_with_metadata() to create a forward dataset, alongside
reverse_sequence_if_needed(), reverse_quality_if_needed(), and reverse_probabilities_if_needed().
Example:
Forward sequence, with indices of Cs in CpGs numbered:
C C C A G G C G G C G G C G A C C G A
            7     10    13      17
(length = 19, locations = &quot;7,10,13,17&quot;, CpGs = 7-8, 10-11, 13-14, 17-18)
Reverse sequence, with indices of C in CpGs numbered:
T C G G T C G C C G C C G C C T G G G
  2       6     9     12
(length = 19, locations = &quot;2,6,9,12&quot;, CpGs = 2-3, 6-7, 9-10, 12-13)
As CG reverse-complements to itself, each CpG site has a 1:1 correspondence with
a CpG site in the reverse strand. Many methylation calling models assess C-methylation
at the C of each CpG. To map the locations from C to C, we take 19 - &amp;lt;index&amp;gt; such that
&quot;7,10,13,17&quot; becomes &quot;12,9,6,2&quot; and &quot;2,6,9,12&quot; becomes &quot;17,13,10,7&quot;.
The symmetry of CpGs means mapping from C to C is also symmetric.
This is achieved by setting offset = 1, as mapping C to C involves shifting position by 1.
Conversely, to map the locations from C to G (i.e. preserving the actual location of each
modification, which is required if assessed locations are non-symmetric/don't reverse-complement
to themselves like CpGs do), we take 20 - &amp;lt;index&amp;gt; such that
&quot;7,10,13,17&quot; becomes &quot;13,10,7,3&quot; i.e. the indices of the Gs in CpGs in the reverse
sequence. Likewise &quot;2,6,9,12&quot; becomes &quot;18,14,11,8&quot; i.e. the indices of the Gs in CpGs in
the forward sequence.
This is achieved by setting offset = 0, as mapping C to G preserves the actual original position
at which each modification was assessed, but changes the base to its complement.
In general, new locations are calculated as (&amp;lt;length&amp;gt; + 1 - &amp;lt;offset&amp;gt;) - &amp;lt;index&amp;gt;.
Of course, output locations are reversed before returning so that they all
return in ascending order, as is standard for all location vectors/strings.
If wanting to write reversed sequences to FASTQ via write_modified_fastq(), locations
must be symmetric (e.g. CpG) and offset must be set to 1. Asymmetric locations are impossible
to write to modified FASTQ once reversed because then e.g. cytosine methylation will be assessed
at guanines, which SAMtools can't account for. Symmetrically reversing CpGs via offset = 1 is
the only way to fix this."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggDNAvis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.3.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ejade42/ggDNAvis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reverse modification locations if needed (<code><a href="merge_methylation_with_metadata.html">merge_methylation_with_metadata()</a></code> helper)</h1>
      <small class="dont-index">Source: <a href="https://github.com/ejade42/ggDNAvis/blob/main/R/merge_with_metadata.R" class="external-link"><code>R/merge_with_metadata.R</code></a></small>
      <div class="d-none name"><code>reverse_locations_if_needed.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function takes a vector of condensed modification locations/indices (e.g.
<code>c("3,6,9,12", "1,4,7,10")</code>), a vector of directions (which must all be either
<code>"forward"</code> or <code>"reverse"</code>, <em>not</em> case-sensitive), and a vector of sequence lengths
(integers).<br><br>
Returns a vector of condensed locations where reads that were originally forward
are unchanged, and reads that were originally reverse are flipped to now be forward.<br><br>
Optionally, a numerical offset can be set. If this is left at <code>0</code> (the default value),
then a CpG assessed for methylation would be reverse-complemented to a CG with the
modification information ascribed to the G (as the G is at the location where the actual
modified C was on the other strand). However, setting the offset to <code>1</code> would shift all
of the modification indices by 1 such that the modification is now ascribed to the C of the
reverse-strand CG. This is beneficial for visualising the modifications as it ensures consistency
between originally-forward and originally-reverse strands by making the modification score associated
with each CpG site always be located at the C, but may be misleading for quantitative analysis.
Setting the offset to anything other than <code>0</code> or <code>1</code> should work but may be biologically misleading,
so produces a warning.<br><br>
Called by <code><a href="merge_methylation_with_metadata.html">merge_methylation_with_metadata()</a></code> to create a forward dataset, alongside
<code><a href="reverse_sequence_if_needed.html">reverse_sequence_if_needed()</a></code>, <code><a href="reverse_quality_if_needed.html">reverse_quality_if_needed()</a></code>, and <code><a href="reverse_probabilities_if_needed.html">reverse_probabilities_if_needed()</a></code>.<br><br><strong>Example:</strong><br><br>
Forward sequence, with indices of Cs in CpGs numbered:<br></p><div class="sourceCode"><pre><code>
C C C A G G C G G C G G C G A C C G A
            7     10    13      17
</code></pre></div><p>(length = 19, locations = <code>"7,10,13,17"</code>, CpGs = 7-8, 10-11, 13-14, 17-18)<br><br>
Reverse sequence, with indices of C in CpGs numbered:<br></p><div class="sourceCode"><pre><code>
T C G G T C G C C G C C G C C T G G G
  2       6     9     12
</code></pre></div><p>(length = 19, locations = <code>"2,6,9,12"</code>, CpGs = 2-3, 6-7, 9-10, 12-13)<br><br>
As CG reverse-complements to itself, each CpG site has a 1:1 correspondence with
a CpG site in the reverse strand. Many methylation calling models assess C-methylation
at the C of each CpG. To map the locations from C to C, we take <code>19 - &lt;index&gt;</code> such that
<code>"7,10,13,17"</code> becomes <code>"12,9,6,2"</code> and <code>"2,6,9,12"</code> becomes <code>"17,13,10,7"</code>.
The symmetry of CpGs means mapping from C to C is also symmetric.
<em>This is achieved by setting <strong><code>offset = 1</code></strong>, as mapping C to C involves shifting position by 1.</em><br><br>
Conversely, to map the locations from C to G (i.e. preserving the actual location of each
modification, which is required if assessed locations are non-symmetric/don't reverse-complement
to themselves like CpGs do), we take <code>20 - &lt;index&gt;</code> such that
<code>"7,10,13,17"</code> becomes <code>"13,10,7,3"</code> i.e. the indices of the Gs in CpGs in the reverse
sequence. Likewise <code>"2,6,9,12"</code> becomes <code>"18,14,11,8"</code> i.e. the indices of the Gs in CpGs in
the forward sequence.
<em>This is achieved by setting <strong><code>offset = 0</code></strong>, as mapping C to G preserves the actual original position
at which each modification was assessed, but changes the base to its complement.</em><br><br>
In general, new locations are calculated as <code>(&lt;length&gt; + 1 - &lt;offset&gt;) - &lt;index&gt;</code>.
Of course, output locations are reversed before returning so that they all
return in ascending order, as is standard for all location vectors/strings.<br><br>
If wanting to write reversed sequences to FASTQ via <code><a href="write_modified_fastq.html">write_modified_fastq()</a></code>, locations
must be symmetric (e.g. CpG) and offset must be set to 1. Asymmetric locations are impossible
to write to modified FASTQ once reversed because then e.g. cytosine methylation will be assessed
at guanines, which SAMtools can't account for. Symmetrically reversing CpGs via <code>offset = 1</code> is
the only way to fix this.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">reverse_locations_if_needed</span><span class="op">(</span></span>
<span>  <span class="va">locations_vector</span>,</span>
<span>  <span class="va">direction_vector</span>,</span>
<span>  <span class="va">length_vector</span>,</span>
<span>  offset <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-locations-vector">locations_vector<a class="anchor" aria-label="anchor" href="#arg-locations-vector"></a></dt>
<dd><p><code>character vector</code>. The locations to be reversed for each sequence/read. Each read should have one character value, representing a comma-separated list of indices at which modification was assessed along the read e.g. <code>"3,6,9,12"</code> for all the <code>Cs</code> in <code>GGCGGCGGCGGC</code>.<br><br> These comma-separated characters/strings can be produced from numeric vectors via <code><a href="vector_to_string.html">vector_to_string()</a></code> and converted back to vectors via <code><a href="string_to_vector.html">string_to_vector()</a></code>.</p></dd>


<dt id="arg-direction-vector">direction_vector<a class="anchor" aria-label="anchor" href="#arg-direction-vector"></a></dt>
<dd><p><code>character vector</code>. Whether each sequence is forward or reverse. Must contain only <code>"forward"</code> and <code>"reverse"</code>, but is not case sensitive. Must be the same length as <code>locations_vector</code> and <code>length_vector</code>.</p></dd>


<dt id="arg-length-vector">length_vector<a class="anchor" aria-label="anchor" href="#arg-length-vector"></a></dt>
<dd><p><code>integer vector</code>. The length of each sequence. Needed for reversing locations as locations are stored relative to the start of the read i.e. relative to the end of the reverse read. Must be the same length as <code>locations_vector</code> and <code>direction_vector</code>.</p></dd>


<dt id="arg-offset">offset<a class="anchor" aria-label="anchor" href="#arg-offset"></a></dt>
<dd><p><code>integer</code>. How much locations should be shifted by. Defaults to <code>0</code>. This is important because if a CpG is assessed for methylation at the C, then reverse complementing it will give a methylation score at the G on the reverse-complemented strand. This is the most biologically accurate, but for visualising methylation it may be desired to shift the locations by <code>1</code> i.e. to correspond with the C in the reverse-complemented CpG rather than the G, which allows for consistent visualisation between forward and reverse strands. Setting (integer) values other than <code>0</code> or <code>1</code> will work, but may be biologically misleading so it is not recommended.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>character vector</code>. A vector of all forward versions of the input locations vector.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">reverse_locations_if_needed</span><span class="op">(</span></span></span>
<span class="r-in"><span>    locations_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"7,10,13,17"</span>, <span class="st">"2,6,9,12"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    direction_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"forward"</span>, <span class="st">"reverse"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    length_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">19</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    offset <span class="op">=</span> <span class="fl">0</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "7,10,13,17" "8,11,14,18"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">reverse_locations_if_needed</span><span class="op">(</span></span></span>
<span class="r-in"><span>    locations_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"7,10,13,17"</span>, <span class="st">"2,6,9,12"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    direction_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"forward"</span>, <span class="st">"reverse"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    length_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">19</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    offset <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "7,10,13,17" "7,10,13,17"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evelyn Jade.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

