<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Convert absolute index locations to MM tag (write_modified_fastq() helper) — convert_locations_to_MM_vector • ggDNAvis</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Convert absolute index locations to MM tag (write_modified_fastq() helper) — convert_locations_to_MM_vector"><meta name="description" content="This function takes a vector of modified base locations as absolute indices
(i.e. a 1 would mean the first base in the sequence has been assessed for
modification; a 15 would mean the 15th base has), and converts it to a vector
in the format of the SAM/BAM MM tags. The MM tag defines a particular target base (e.g.
C for methylation), and then stores the number of skipped instances of that base
between sites where modification was assessed. In practice, this often means counting the
number of non-CpG Cs in between CpG Cs. In a GGC repeat, this should be a bunch of 0s
as every C is in a CpG, but unique sequence will have many non-CpG Cs.
This function is reversed by convert_MM_vector_to_locations()."><meta property="og:description" content="This function takes a vector of modified base locations as absolute indices
(i.e. a 1 would mean the first base in the sequence has been assessed for
modification; a 15 would mean the 15th base has), and converts it to a vector
in the format of the SAM/BAM MM tags. The MM tag defines a particular target base (e.g.
C for methylation), and then stores the number of skipped instances of that base
between sites where modification was assessed. In practice, this often means counting the
number of non-CpG Cs in between CpG Cs. In a GGC repeat, this should be a bunch of 0s
as every C is in a CpG, but unique sequence will have many non-CpG Cs.
This function is reversed by convert_MM_vector_to_locations()."><meta property="og:image" content="https://ejade42.github.io/ggDNAvis/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggDNAvis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.2.9009</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-interactive-app" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Interactive App</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-interactive-app"><li><a class="dropdown-item" href="../articles/interactive_app.html">Embedded</a></li>
    <li><a class="external-link dropdown-item" href="https://ejade42-ggdnavis.share.connect.posit.cloud/" target="_blank">Fullscreen</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-version" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Version</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-version"><li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/">Latest release</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/dev/">Latest dev</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.3.2/">v0.3.2</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.3.1/">v0.3.1</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.3.0/">v0.3.0</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.2.1/">v0.2.1</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.2.0/">v0.2.0</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ejade42/ggDNAvis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Convert absolute index locations to MM tag (<code><a href="write_modified_fastq.html">write_modified_fastq()</a></code> helper)</h1>
      <small class="dont-index">Source: <a href="https://github.com/ejade42/ggDNAvis/blob/main/R/parse_methylation_from_fastq.R" class="external-link"><code>R/parse_methylation_from_fastq.R</code></a></small>
      <div class="d-none name"><code>convert_locations_to_MM_vector.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function takes a vector of modified base locations as absolute indices
(i.e. a <code>1</code> would mean the first base in the sequence has been assessed for
modification; a <code>15</code> would mean the 15th base has), and converts it to a vector
in the format of the SAM/BAM MM tags. The MM tag defines a particular target base (e.g.
<code>C</code> for methylation), and then stores the number of skipped instances of that base
between sites where modification was assessed. In practice, this often means counting the
number of non-CpG <code>C</code>s in between CpG <code>C</code>s. In a <code>GGC</code> repeat, this should be a bunch of <code>0</code>s
as every <code>C</code> is in a CpG, but unique sequence will have many non-CpG <code>C</code>s.<br><br>
This function is reversed by <code><a href="convert_MM_vector_to_locations.html">convert_MM_vector_to_locations()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">convert_locations_to_MM_vector</span><span class="op">(</span><span class="va">sequence</span>, <span class="va">locations</span>, target_base <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sequence">sequence<a class="anchor" aria-label="anchor" href="#arg-sequence"></a></dt>
<dd><p><code>character</code>. The DNA sequence about which the methylation information is being processed.</p></dd>


<dt id="arg-locations">locations<a class="anchor" aria-label="anchor" href="#arg-locations"></a></dt>
<dd><p><code>integer vector</code>. All of the base indices at which methylation/modification information was processed. Must all be instances of the target base.</p></dd>


<dt id="arg-target-base">target_base<a class="anchor" aria-label="anchor" href="#arg-target-base"></a></dt>
<dd><p><code>character</code>. The base type that has been assessed or skipped (defaults to <code>"C"</code>).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>integer vector</code>. A component of a SAM MM tag, representing the number of skipped target bases in between each assessed base.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">convert_locations_to_MM_vector</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"GGCGGCGGCGGC"</span>,</span></span>
<span class="r-in"><span>    locations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">9</span>, <span class="fl">12</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    target_base <span class="op">=</span> <span class="st">"C"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0 0 0 0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">convert_locations_to_MM_vector</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"GGCGGCGGCGGC"</span>,</span></span>
<span class="r-in"><span>    locations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">7</span>, <span class="fl">10</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    target_base <span class="op">=</span> <span class="st">"G"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0 1 1 1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">convert_locations_to_MM_vector</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"GGCGGCGGCGGC"</span>,</span></span>
<span class="r-in"><span>    locations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">10</span>, <span class="fl">11</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    target_base <span class="op">=</span> <span class="st">"G"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0 0 0 0 0 0 0 0</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evelyn Jade.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

