<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Write sequence and quality information to FASTQ — write_fastq • ggDNAvis</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Write sequence and quality information to FASTQ — write_fastq"><meta name="description" content="This function simply writes a FASTQ file from a dataframe containing
columns for read ID, sequence, and quality scores.
See fastq_quality_scores for an explanation of quality.
Said dataframe can be produced from FASTQ via read_fastq().
To read/write a modified FASTQ containing modification information
(SAM/BAM MM and ML tags) in the header lines, use
read_modified_fastq() and write_modified_fastq()."><meta property="og:description" content="This function simply writes a FASTQ file from a dataframe containing
columns for read ID, sequence, and quality scores.
See fastq_quality_scores for an explanation of quality.
Said dataframe can be produced from FASTQ via read_fastq().
To read/write a modified FASTQ containing modification information
(SAM/BAM MM and ML tags) in the header lines, use
read_modified_fastq() and write_modified_fastq()."><meta property="og:image" content="https://ejade42.github.io/ggDNAvis/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggDNAvis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.2.9001</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-version" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Version</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-version"><li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/">Latest release</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/dev/">Latest dev</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.3.2/">v0.3.2</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.3.1/">v0.3.1</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.3.0/">v0.3.0</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.2.1/">v0.2.1</a></li>
    <li><a class="dropdown-item" href="https://ejade42.github.io/ggDNAvis/v0.2.0/">v0.2.0</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ejade42/ggDNAvis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Write sequence and quality information to FASTQ</h1>
      <small class="dont-index">Source: <a href="https://github.com/ejade42/ggDNAvis/blob/main/R/parse_methylation_from_fastq.R" class="external-link"><code>R/parse_methylation_from_fastq.R</code></a></small>
      <div class="d-none name"><code>write_fastq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function simply writes a FASTQ file from a dataframe containing
columns for read ID, sequence, and quality scores.<br><br>
See <code><a href="fastq_quality_scores.html">fastq_quality_scores</a></code> for an explanation of quality.<br><br>
Said dataframe can be produced from FASTQ via <code><a href="read_fastq.html">read_fastq()</a></code>.
To read/write a modified FASTQ containing modification information
(SAM/BAM MM and ML tags) in the header lines, use
<code><a href="read_modified_fastq.html">read_modified_fastq()</a></code> and <code><a href="write_modified_fastq.html">write_modified_fastq()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">write_fastq</span><span class="op">(</span></span>
<span>  <span class="va">dataframe</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  read_id_colname <span class="op">=</span> <span class="st">"read"</span>,</span>
<span>  sequence_colname <span class="op">=</span> <span class="st">"sequence"</span>,</span>
<span>  quality_colname <span class="op">=</span> <span class="st">"quality"</span>,</span>
<span>  return <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dataframe">dataframe<a class="anchor" aria-label="anchor" href="#arg-dataframe"></a></dt>
<dd><p>Dataframe containing sequence information to write back to FASTQ. Must have columns for unique read ID and DNA sequence. Should also have a column for quality, unless wanting to fill in qualities with <code>"B"</code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p><code>character</code>. File to write the FASTQ to. Recommended to end with <code>.fastq</code> (warns but works if not). If set to <code>NA</code> (default), no file will be output, which may be useful for testing/debugging.</p></dd>


<dt id="arg-read-id-colname">read_id_colname<a class="anchor" aria-label="anchor" href="#arg-read-id-colname"></a></dt>
<dd><p><code>character</code>. The name of the column within the dataframe that contains the unique ID for each read. Defaults to <code>"read"</code>.</p></dd>


<dt id="arg-sequence-colname">sequence_colname<a class="anchor" aria-label="anchor" href="#arg-sequence-colname"></a></dt>
<dd><p><code>character</code>. The name of the column within the dataframe that contains the DNA sequence for each read. Defaults to <code>"sequence"</code>.<br><br> The values within this column must be DNA sequences e.g. <code>"GGCGGC"</code>.</p></dd>


<dt id="arg-quality-colname">quality_colname<a class="anchor" aria-label="anchor" href="#arg-quality-colname"></a></dt>
<dd><p><code>character</code>. The name of the column within the dataframe that contains the FASTQ quality scores for each read. Defaults to <code>"quality"</code>. If scores are not known, can be set to <code>NA</code> to fill in quality with <code>"B"</code>.<br><br> If not <code>NA</code>, must correspond to a column where the values are the FASTQ quality scores e.g. <code>"$12\"&gt;/2C;4:9F8:816E,6C3*,"</code> - see <code><a href="fastq_quality_scores.html">fastq_quality_scores</a></code>.</p></dd>


<dt id="arg-return">return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p><code>logical</code>. Boolean specifying whether this function should return the FASTQ (as a character vector of each line in the FASTQ), otherwise it will return <code>invisible(NULL)</code>. Defaults to <code>FALSE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>character vector</code>. The resulting FASTQ file as a character vector of its constituent lines (or <code>invisible(NULL)</code> if <code>return</code> is <code>FALSE</code>). This is probably mostly useful for debugging, as setting <code>filename</code> within this function directly writes to FASTQ via <code><a href="https://rdrr.io/r/base/writeLines.html" class="external-link">writeLines()</a></code>. Therefore, defaults to returning <code>invisible(NULL)</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Write to FASTQ (using filename = NA, return = FALSE</span></span></span>
<span class="r-in"><span><span class="co">## to view as char vector rather than writing to file)</span></span></span>
<span class="r-in"><span><span class="fu">write_fastq</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">example_many_sequences</span>,</span></span>
<span class="r-in"><span>    filename <span class="op">=</span> <span class="cn">NA</span>,</span></span>
<span class="r-in"><span>    read_id_colname <span class="op">=</span> <span class="st">"read"</span>,</span></span>
<span class="r-in"><span>    sequence_colname <span class="op">=</span> <span class="st">"sequence"</span>,</span></span>
<span class="r-in"><span>    quality_colname <span class="op">=</span> <span class="st">"quality"</span>,</span></span>
<span class="r-in"><span>    return <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in write_fastq(example_many_sequences, filename = NA, read_id_colname = "read",     sequence_colname = "sequence", quality_colname = "quality",     return = TRUE):</span> '...' used in an incorrect context</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## quality_colname = NA fills in quality with "B"</span></span></span>
<span class="r-in"><span><span class="fu">write_fastq</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">example_many_sequences</span>,</span></span>
<span class="r-in"><span>    filename <span class="op">=</span> <span class="cn">NA</span>,</span></span>
<span class="r-in"><span>    read_id_colname <span class="op">=</span> <span class="st">"read"</span>,</span></span>
<span class="r-in"><span>    sequence_colname <span class="op">=</span> <span class="st">"sequence"</span>,</span></span>
<span class="r-in"><span>    quality_colname <span class="op">=</span> <span class="cn">NA</span>,</span></span>
<span class="r-in"><span>    return <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in write_fastq(example_many_sequences, filename = NA, read_id_colname = "read",     sequence_colname = "sequence", quality_colname = NA, return = TRUE):</span> '...' used in an incorrect context</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Evelyn Jade.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

